### PR TITLE
Refactor SwiftPMConfig

### PR DESCRIPTION
**Pull request by mattt**
As discussed in #2995, the `SwiftPMConfig` type presents some API design challenges:

> The `SwiftPMConfig` type is responsible for getting and setting the mirrored URLs for dependencies. For example, if a package has a direct or transitive dependency on `https://github.com/apple/swift-argument-parser.git`, you can use a mirror to instead point all references to that package to a fork, say `https://github.com/mattt/swift-argument-parser`. This functionality was discussed on the Swift Forums [in this thread](https://forums.swift.org/t/dependency-mirroring-and-forking/13902).
> 
> The name `SwiftPMConfig` is vague, and makes it difficult to understand the significance of `config` properties and parameters that are passed around among the `PackageGraph`, `SourceControl`, and `Workspace` modules. In other contexts, the word "config" is used to describe `pkgConfig`, build configuration, or `xcconfig` files.
> 
> [This comment](https://github.com/apple/swift-package-manager/blob/7afc6e22bb7b957b7d212fea31cf0f9132513bc0/Sources/SourceControl/SwiftPMConfig.swift#L16) indicates that the name was chosen in anticipation of the type being used to manage other concerns.

In the feedback on that PR, @neonichu recommended making mirrors a subtype / subproperty on `SwiftPMConfig`, thereby not closing off the opportunity to extend `SwiftPMConfig` to other purposes at some point in the future.

This PR follows that guidance — to great effect, I believe. 

First, I pulled in some of the other changes from #2995, like reimplementing `SwiftPMConfig` to more closely match `PinsStore` (51c282f4073cf5cd81df36a65ba49e23e3cc6469) and adding test coverage (9089f5d7f2721f8c1b820b3740a469f40352b1a5). I also added documentation (dbecb53ceeafba1f14dd27c2c5ca5275dc46bd08), and in the process saw an opportunity to improve clarity by renaming `mirroredURL(forURL:)` to `effectiveURL(forURL:)` (da0eac720b50efab22108524ffda328342d47a59). 

From there, I made a `DependencyMirrors` type, which conforms to `Collection` and acts as a proxy for getting and setting mirrors, and re-routed all existing calls through that (48f77f4e1a32a0269bd01d5ba516996e36d97ccd). Along the way, I noticed that the only method being called on mirrors outside of the `Command` module was that renamed `effectiveURL(forURL:)` method from before. Thanks to that, I was able to find a solution to the problem of all of this API design awkwardness: 

1. Move `SwiftPMConfig` to the `Workspace` module
2. Make `DependencyMirrors` a standalone type in the `PackageGraph` module
 
Thanks to these separation of concerns, we can now perform all of our dependency resolution tasks without needing to know about `SwiftPMConfig` at all. Each `config` parameter or property can be (and was) replaced by a corresponding one named `mirrors`. And should `SwiftPMConfig` ever expand its scope, we can always add another property / argument to the relevant call sites as needed.

The one notable thing I punted on so far with this PR was renaming `SwiftPMConfig`. It doesn't _need_ a new name, but its repositioning in the project gives us an occasion to revisit that decision. I think `Workspace.Configuration` would be an improvement, but I'm interested to hear any other ideas.

### CODE REVIEW

**Comment by mattt** [comment]
@swift-ci please smoke test

**Comment by mattt** [comment]
@swift-ci please smoke test

**Comment by mattt** [comment]
CI is failing for the same reason discussed in https://github.com/apple/swift-package-manager/pull/2995#issuecomment-715538725.

```
/home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-main/sourcekit-lsp/Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift:96:15: error: call can throw but is not marked with 'try'
15:51:34       config: SwiftPMConfig(path: packageRoot.appending(components: ".swiftpm", "config"), fs: fileSystem),
15:51:34               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15:51:34 /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-main/sourcekit-lsp/Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift:96:15: note: did you mean to use 'try'?
15:51:34       config: SwiftPMConfig(path: packageRoot.appending(components: ".swiftpm", "config"), fs: fileSystem),
15:51:34               ^
15:51:34               try 
15:51:34 /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-main/sourcekit-lsp/Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift:96:15: note: did you mean to handle error as optional value?
15:51:34       config: SwiftPMConfig(path: packageRoot.appending(components: ".swiftpm", "config"), fs: fileSystem),
15:51:34               ^
15:51:34               try? 
15:51:34 /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-main/sourcekit-lsp/Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift:96:15: note: did you mean to disable error propagation?
15:51:34       config: SwiftPMConfig(path: packageRoot.appending(components: ".swiftpm", "config"), fs: fileSystem),
15:51:34               ^
15:51:34               try! 
```

A few options for resolving this issue:

1. Pinning SwiftPM dependency in sourcekit-lsp to semantic version
2. Coordinating simultaneous PR on sourcekit-lsp with corresponding API changes
3. Reverting to previous API

Of these, I think 1 is the best (and probably what we should be doing anyway; dependencies on branches are fragile). 2 could work but is risky and will almost certainly break the build temporarily. Otherwise, 3 is always an option.

@neonichu @abertelrud Any preference among these three options? (Or any other ideas?)

**Comment by MaxDesiatov** [comment]
> 2. Coordinating simultaneous PR on sourcekit-lsp with corresponding API changes

If I understand correctly, this is how it's usually done here. You open multiple PRs across different repositories and trigger CI with PR conditions so that it pulls exact PRs across multiple repos to test with. When everything passes, you merge it all at once so that the amount of time that the build is broken is minimized.

**Comment by tomerd** [comment]
@swift-ci please test

**Comment by tomerd** [comment]
I like this change allot. I am also +1 on renaming to `Workspace.Configuration` which is easier to reason about. 

Incidentally, some of the work that will follow up on https://github.com/apple/swift-package-manager/pull/3007 includes some similar concepts so we would want to harmonize them as these things land

File: Sources/PackageGraph/DependencyMirrors.swift [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    }
+}
+
+extension DependencyMirrors: ExpressibleByDictionaryLiteral {
+    public convenience init(dictionaryLiteral elements: (String, String)...) {
+        self.init(elements.map { Mirror(original: $0.0, mirror: $0.1) })
+    }
+}
+
+extension DependencyMirrors: JSONMappable, JSONSerializable {
```
Comment (tomerd) [inline Sources/PackageGraph/DependencyMirrors.swift:91]: does this need to be using Codable instead?

File: Sources/Workspace/SwiftPMConfig.swift [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+           let configFile = self.configFile
+        {
+            return try fileSystem.removeFileTree(configFile)
+        }
+
+        try persistence.saveState(self)
+    }
+}
+
+extension SwiftPMConfig: JSONSerializable {
```
Comment (tomerd) [inline Sources/Workspace/SwiftPMConfig.swift:92]: Codable?

File: Sources/PackageGraph/DependencyMirrors.swift (hunk 1 above)
Comment (mattt) [reply Sources/PackageGraph/DependencyMirrors.swift:91]: Both here and [here](https://github.com/apple/swift-package-manager/pull/3002#discussion_r515305732), I'm deferring to the original implementation, which uses these pre-Codable JSON serialization protocols. I'd be just as happy to switch to Codable, but I didn't know what our current disposition is (i.e. _stare decisis_ vs. "clean up as we go along").

**Comment by mattt** [comment]
> Incidentally, some of the work that will follow up on #3007 includes some similar concepts so we would want to harmonize them as these things land

Oh, interesting. Do you have a link you can share with more information about this feature?

**Comment by tomerd** [comment]
@swift-ci please test

File: Sources/PackageGraph/DependencyMirrors.swift (hunk 1 above)
Comment (tomerd) [reply Sources/PackageGraph/DependencyMirrors.swift:91]: the diff made me think this is new code (where we would prefer Codable), but if its just moving existing code around I think we can focus the PR on what its actually trying to achieve.

**Comment by tomerd** [comment]
> Oh, interesting. Do you have a link you can share with more information about this feature?

see https://forums.swift.org/t/package-feeds/41522, the "Configuration file" section is the relevant one. I can see some of the mirror and/or mirror like configuration being beyond workspace.

File: Sources/PackageGraph/DependencyMirrors.swift (hunk 1 above)
Comment (abertelrud) [reply Sources/PackageGraph/DependencyMirrors.swift:91]: Agreed; separately we should move everything to Codable, but as separately testable PRs.

**Comment by abertelrud** [comment]
Thanks @mattt, this is great.  Refactoring so that the configuration type (whatever it's called) is just a holder of more specific types that the logic then reasons about is good.  I also think that `Workspace.Configuration` is a good name.

Regarding the revlock issue:  we _really_ want to get to the point where the various packages here are using semantic versioning, but for the moment, with the existing branch-based dependencies, I believe @MaxDesiatov is right that opening separate PRs is how it's usually been done.  But in my view we need to quickly get to the point where we're using semantic versioning for the dependencies, even if that means having to increment the versions often as we don't yet have a stable API.

**Review by abertelrud — APPROVED** [review]

**Review by neonichu — APPROVED** [review]

**Comment by mattt** [comment]
Thanks for reviewing, @abertelrud & @tomerd. I just committed a2876c7, which takes that final step to rename `SwiftPMConfig` to `Workspace.Configuration`.

I also opened https://github.com/apple/sourcekit-lsp/pull/344 which has the corresponding changes in sourcekit-lsp. There may be other Apple OSS projects that require an update; I'm only aware of this one because CI stopped there.

**Comment by neonichu** [comment]
apple/sourcekit-lsp#344
@swift-ci please smoke test

**Comment by benlangmuir** [comment]
> Regarding the revlock issue: we really want to get to the point where the various packages here are using semantic versioning, but for the moment, with the existing branch-based dependencies, I believe @MaxDesiatov is right that opening separate PRs is how it's usually been done. But in my view we need to quickly get to the point where we're using semantic versioning for the dependencies, even if that means having to increment the versions often as we don't yet have a stable API.

@abertelrud what is your plan here?  Is there work happening to guarantee swiftpm is stable with respect to build directory layout and intermediate file formats?  Rev-lock is currently the only guarantee of compatibility between sourcekit-lsp and the package manager in the toolchain.

**Comment by abertelrud** [comment]
@benlangmuir  This is currently a goal, not a plan.  The plan would be to first define what's part of the API for libSwiftPM (such as the file formats and the semantics you mention), and then to start versioning it.  We would absolutely need to coordinate that with SourceKit-LSP.  My comments were just pointing out that as libSwiftPM gains more clients, it becomes less feasible to manage sets of interlocking changes.

**Pull request merged**